### PR TITLE
make data hashable

### PR DESCRIPTION
**Pull request by LinguList**
It is crucial in many situations to be able to hash a typed sequence, so this code provides, hopefully, all we need. Now, a typedsequence can also function as a key in a dictionary.

### CODE REVIEW

File: src/linse/typedsequence.py [hunk 1]
```diff
@@ -37,6 +37,15 @@ def write(item):
     def __str__(self):
         return ' '.join([self.__class__.write(x) for x in self])
 
+    def __repr__(self):
+        return repr(str(self))
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __eq__(self):
```
Comment (xrotwang) [inline src/linse/typedsequence.py]: That doesn't look right. The signature ought to be `def __eq__(self, other)`.

**Review by xrotwang — CHANGES_REQUESTED** [review]
We also need tests :)

**Comment by LinguList** [comment]
Yes, I found in fact that we must add more tests, I'll try and add them 
later!

**Comment by LinguList** [comment]
Problems should be addressed now. The modification also fixes a problem resulting from appending something to a word, which resulted in not appending something tot he morphemes before. This is in my opinion fixed now.

**Comment by xrotwang** [comment]
One question: Do we need a custom `__eq__` implementation? Isn't comparing the underlying sequences ok (or better)?

**Comment by LinguList** [comment]
The eq-function reduces to the string representation, but I consider this as a feature rather than a problem. First, the string representation in CSV and other formats will be anyway the most clear representation of a word or a list of integers. Second, we cannot hash a word if we do not define an eq in this way (at least it throws errors and I read one must make sure to be consistent with eq when defining a hash).

**Comment by LinguList** [comment]
It also means -- which I consider a feature -- that one can populare a dictionary with words in Python and later access words by their string representation.

**Comment by LinguList** [comment]
Minimal example:

```python
w = Word("a + b")
d = {w: 1}
d["a + b"]
```
This yields a key error if we do not default to string representation when comparing in eq.

**Comment by xrotwang** [comment]
I've just seen the test, that uses a literal string to retrieve something from a dictionary. I consider this a mis-feature, though. It's basically circumventing - and actually preventing - any useful type checking. E.g. the type hint for a dictionary keyed by `Word`s would now need to be `typing.Dict[typing.Union[str, Word], ...]` rather than just `typing.Dict[Word, ...]`.

I just fixed the failing test (after removing the `__eq__` method) by wrapping the literal string into `Word(...)`.

Having objects of one type behave like a different type in some circumstances does not only confuse a type checker. So I'd opt to remove it.

**Comment by LinguList** [comment]
You are right. It is also not a real problem to populate a dictionary with strings. We could then even get rid of the hash itself.

**Comment by xrotwang** [comment]
> Minimal example:
> 
> ```python
> w = Word("a + b")
> d = {w: 1}
> d["a + b"]
> ```
> 
> This yields a key error if we do not default to string representation when comparing in eq.

But this works:
```
>>> d[Word("a + b")]
```
and seems simple enough to me. I find the following weirder:
```python
>>> from linse.typedsequence import Word
>>> w = Word('a + b')
>>> 'a + b' == w
True
>>> 'a + b'.morphemes == w.morphemes
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'str' object has no attribute 'morphemes'
```

**Comment by xrotwang** [comment]
> You are right. It is also not a real problem to populate a dictionary with strings. We could then even get rid of the hash itself.

Making the thing hashable might be useful, see
```python
>>> {Word('a + b'), Word('a + c'), Word('a + b')}
{'a + c', 'a + b'}
```

**Review by xrotwang — APPROVED** [review]

**Comment by LinguList** [comment]
Yes! I thought of dictionaries for now, but the SET feature is very cool for our purpose, even more important.

**Pull request merged**